python/ray/experimental/data/impl/transform_util.py

```diff
@@ -0,0 +1,48 @@
+import itertools
+try:
+    import pyarrow as pa
+except ImportError:
+    pa = None
+try:
+    import pandas as pd
+except ImportError:
+    pd = None
+
+from functools import reduce
+from typing import List, Union, Iterable, TypeVar
+
+from ray.experimental.data.impl.arrow_block import (ArrowBlockBuilder,
+                                                    ArrowBlock)
+from ray.experimental.data.impl.block import Block, ListBlock
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _import_check():
+    if pa is None:
+        raise ImportError("Run `pip install pyarrow` for Arrow support")
+    if pd is None:
+        raise ImportError("Run `pip install pandas` for Pandas support")
+
+
+def concat_batches(batches: Iterable[Union["pd.DataFrame", List, "pa.Table"]]
```

Comment:
Move any needed logic into DelegatingArrowBlockBuilder.

Reply:
It seems to be hard to move there unless we use `add` directly. I moved it to the transform function and just use add_block instead.